Add support for Aqara external temperature sensor

Antoninj opened the pull request:
## Proposed change
Support for external temperature sensor.

## Additional information

This is a copy of https://github.com/zigpy/zha-device-handlers/pull/2802 + the missing unit tests

## Checklist

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works



Related: #3585

Antoninj committed:
update existing quirk based on https://github.com/zigpy/zha-device-handlers/pull/2802

Antoninj committed:
add test

Antoninj committed:
fix README.md

Antoninj committed:
fix tests

Antoninj committed:
Merge branch 'dev' into aqara_external_temperature_sensor

Antoninj committed:
fix tests

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

Antoninj committed:
fix linting issues

Antoninj committed:
Merge remote-tracking branch 'origin/aqara_external_temperature_sensor' into aqara_external_temperature_sensor

Antoninj committed:
renaming

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -388,6 +395,8 @@ class AqaraThermostatSpecificCluster(XiaomiAqaraE1Cluster):
             SCHEDULE_SETTINGS: ("schedule_settings", ScheduleSettings, True),
             SENSOR: ("sensor", t.uint8_t, True),
             BATTERY_PERCENTAGE: ("battery_percentage", t.uint8_t, True),
+            SENSOR_TEMP: ("sensor_temp", t.uint32_t, True),
+            SENSOR_ATTR: (SENSOR_ATTR_NAME, t.LVBytes, True),
```
All these attributes should really be converted to the new zigpy `AttributeDefs` style. Search `AttributeDefs` in this repo or the zigpy repo for examples.
You don't have to do it in this PR, but it would be nice.

We can also access the name and id then: `AqaraThermostatSpecificCluster.AttributeDefs.sensor_temp.name` or `.id`.

Antoninj replied:
I hear you but I honestly don't have much time to spend on this PR atm and multiple people have been emailing me to track progress on this change since it's a highly requested feature for the quirk.

Elyasin replied:
This is quite some time ago. I would like to help in this case.

I found that this PR is not up to date with the upstream.
In the upstream I see that the AttributeDefs are already defined.
So it is, I believe, straight forward to do this change.

Unfortunately, I don't know how I can collaborate on this PR.

@Antoninj - I believe you can do that quite easily by doing the following:

1. Sync the fork (dev branch) with your the upstream (on GitHub you can do that easily)
2. Pull the updated dev branch to your local machine (e.g. `git checkout dev` and then `git pull`)
3. Now you can rebase the branch `aqara_external_temperature_sensor` onto  dev branch (or merge dev branch into your branch)
4. You will have merge conflicts, which are easily resolved:
  - The test file is new and is not affected
  - The file `thermostat_agl001.py` is already in the new style and you should keep the code of your branch, except the changes of the attributes. These are already in the new style and you need to add those two lines I believe:
```
        sensor_temp: Final = ZCLAttributeDef(
            id=SENSOR_TEMP, type=t.uint32_t, is_manufacturer_specific=True
        )
        sensor_attr: Final = ZCLAttributeDef(
            id=SENSOR_ATTR, type=t.LVBytes, is_manufacturer_specific=True
        )
```

The tests run fine when I tried it.

However, there is the comment of NJLangley, which needs to be addressed. I am trying to look into that. As I am still quite "new" I try to learn und understand what is meant by the comment.

Elyasin replied:
If you feel that you do not have time, I could replicate this PR and do the changes.
Let me know if that is ok for you.

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -402,6 +411,167 @@ def _update_attribute(self, attrid, value):
             )
         super()._update_attribute(attrid, value)
 
+    def aqaraHeader(self, counter: int, params: bytearray, action: int) -> bytearray:
```
This should be snake_case, so `aqara_header`.
Or maybe `wrap_with_aqara_header`? Something like that?

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -402,6 +411,167 @@ def _update_attribute(self, attrid, value):
+                params += temperatureBuf
+
+                attrs = {}
+                attrs[SENSOR_ATTR_NAME] = self.aqaraHeader(0x12, params, 0x05) + params
```
Is `attrs` intentionally set to `{}` again? It's already set at the beginning of the method. Depending on the order of the `attributes` parameter, the others wouldn't be written (although it's rare that this method gets multiple attributes to write at the same time).

 If it's not intentional, this could be rewritten as: 
```python
attrs = {SENSOR_ATTR_NAME: self.aqara_header(0x12, params, 0x05) + params}
```
If not, well, the `attrs = {}` can be removed here.

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -402,6 +411,167 @@ def _update_attribute(self, attrid, value):
+    ) -> list:
+        """Write attributes to device with internal 'attributes' validation."""
+        sensor = bytearray.fromhex("00158d00019d1b98")
+        attrs = {}
```
I think it'd be nice to type this. Should be  `attrs: dict[str | int, Any] = {}` then.

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -402,6 +411,167 @@ def _update_attribute(self, attrid, value):
+                            0x00,
+                            0x00,
+                        ]
+                    )
```
I believe you can just do `params1 += bytes(12)`.

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -402,6 +411,167 @@ def _update_attribute(self, attrid, value):
+                            0x00,
+                            0x00,
+                        ]
+                    )
```
Here as well:
I believe you can just do `params2 += bytes(12)`.

TheJulianJES replied:
Here as well:
I believe you can just do `params2 += bytes(12)`.

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -402,6 +411,167 @@ def _update_attribute(self, attrid, value):
+        for attr, value in attributes.items():
+            # implemented with help from https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/devices/xiaomi.js
+
+            if attr == SENSOR_TEMP:
```
Technically, the `attributes` can either be the id or the name. Generally, it's the id, so this works, but it might be nice to accept both the id and name for `SENSOR` and `SENSOR_TEMP`.

If you were to implement the new-style `AttributeDefs`, you can use `AqaraThermostatSpecificCluster.AttributeDefs.sensor_temp.name` or `AqaraThermostatSpecificCluster.AttributeDefs.sensor_temp.id` to access name/id.

Or you can also use `find_attribute` to get the `ZCLAttributeDef` then. No need to check both id and name then. I think this should work then:
```python
        attrs: dict[str | int, Any] = {}
        for attr, value in attributes.items():
            attr_def = self.find_attribute(attr)
            if attr_def == AqaraThermostatSpecificCluster.AttributeDefs.sensor_temp:
                # do stuff
            else:
                attrs[attr] = value
        return await super().write_attributes(attrs, manufacturer)
```

(This might be streamlined in zigpy in the future.)

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -402,6 +411,167 @@ def _update_attribute(self, attrid, value):
+        self, attributes: dict[str | int, Any], manufacturer: int | None = None
+    ) -> list:
+        """Write attributes to device with internal 'attributes' validation."""
+        sensor = bytearray.fromhex("00158d00019d1b98")
```
I guess we don't need to call this every time here. Probably enough to just move it outside of the method.

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -402,6 +411,167 @@ def _update_attribute(self, attrid, value):
+                        self.aqaraHeader(0x13, params2, 0x04) + params2
+                    )
+
+                    await super().write_attributes(attrs1, manufacturer)
```
So these need to be written in separate calls?

Antoninj replied:
I honestly have no idea, I bootstrapped this PR from existing work from someone else and mostly focused on addding unit tests tbh

MichelFR replied:
They need to be in separate calls otherwise its starts to misbehave

Elyasin replied:
My guess is that two separate calls is necessary, otherwise it would not work correctly.
I conclude that there is nothing to do then.

TheJulianJES commented on README.md:
````diff
@@ -456,7 +456,7 @@ The tests use the [pytest](https://docs.pytest.org/en/latest/) framework.
 
 ```bash
-pip install -r requirements_test_all.txt
+pip install -r requirements_test.txt
````
Thanks for noticing. Let's keep this in a separate PR though.
(Also, `script/setup` should also do this.)

Antoninj replied:
ok, I'll revert no problem

TheJulianJES reviewed (commented):
Thanks for adding the tests. I think this looks pretty good already.

Antoninj committed:
Apply review suggestions for Aqara thermostat external temperature sensor

- Move SENSOR_ID constant outside write_attributes method for efficiency
- Use bytes(12) instead of long lists of zeros for cleaner code
- Improve attribute handling with find_attribute for better robustness
- Add proper type annotations for better code clarity
- Update README.md to reference requirements_test_all.txt
- Run ruff formatting

Antoninj committed:
Rename aqaraHeader to aqara_header for Python naming conventions (code review suggestion)

Antoninj commented:
@TheJulianJES thanks for the detailed code review, I've addressed most of the remarks.

NJLangley commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -402,6 +413,136 @@ def _update_attribute(self, attrid, value):
+                    self._float_to_hex(round(float(value)))[2:]
+                )
+
+                params = SENSOR_ID
```
This seems to break the quirk for me. I can set the TRV into external sensor mode once and set the temp, but all subsequent calls fail and it seems that the max size of the constructed message is exceeded.

It's because the `SENSOR_ID` variable is not copied, so appending more bytes to the `params` variable is modifying `SENSOR_ID`, which means it's then bigger to start with the next time it's run. Changing the line to 
```
params = bytearray(SENSOR_ID)
```
seems to fix it for me. Might want the tests to check the length of the message args for two subsequent calls to verify this?

Elyasin replied:
I tried the suggestion you mentioned and wrote a test for this to avoid regression.

I hope to revceive an answer from @Antoninj in regard to collaborate on this PR or if I re-create the PR with the changes.

NJLangley commented:
I was having some issues with this, as it worked for the first call to set the mode to external sensor and then set the temp, but for all subsequent calls until the ZHA integrations was reloaded then failed. Digging through the stack trace for the error suggests the message exceeded the size of the window used for Zigbee comms. 

The original quirk from #2802 does work, and I've left a comment on the line that needs fixing. Looks like it's not copying the byte array before building the message from it, so it's them broken when it builds the next message

Elyasin commented:
I created a new PR that continues Antonin's work with the review comments taken into account. I added two test cases and did little refactoring, which I believe helps read the code better.

https://github.com/zigpy/zha-device-handlers/pull/4463
